### ISSUE TITLE
Validation data issues?

### ISSUE THREAD

**Issue by ibrahimLearning** [AlexeyAB/darknet#3174]
Although I built yolo from https://pjreddie.com/darknet/, I would like to ask some questions for my clarification.
Would yolo used validation data during training for tuning the model's hyperparameters? 
I searched the "train_detector" function in detector.c, but I cannot find the it reads the validation data.
If it is used the validation data during training, would tell me please where the line of code?
If not use, then how it display Avg IOU,..etc.?
Would you clear me please about the validation and training issues?

**Comment by AlexeyAB** [comment]
> Would yolo used validation data during training for tuning the model's hyperparameters?

No.

Validation set is used only for checking accuracy.

----

> If not use, then how it display Avg IOU,..etc.?

It displays avg IoU for Training set.

**Comment by ibrahimLearning** [comment]
@AlexeyAB Thank you very much for your nice answer. Would you give a suggestion or idea please?

If I want to use validatin data for tuning hyperparameter, how can I do?

**Comment by AlexeyAB** [comment]
Just add `stopbackward=1` in the one of layers (if you want) to freeze all previous layers,
Then set `train=valid.txt` in obj.data file, and conitnure training.

**Comment by ibrahimLearning** [comment]
@AlexeyAB That means, after training the network if I want to tune the network parameter using validation data, I will add stopbackward=1 in the one of layers and use validation data as a training data. Then trained again using the final weight of the trained network(previous). My understanding is OK?

"Just add stopbackward=1 in the one of layers"
Would you clear me please, which layer will I choose to add stopbackward=1 ?

**Comment by AlexeyAB** [comment]
@ibrahimLearning Yes, exactly.

> "Just add stopbackward=1 in the one of layers"
> Would you clear me please, which layer will I choose to add stopbackward=1 ?

2-3 layers before the first `[yolo]` layer

**Comment by ibrahimLearning** [comment]
@AlexeyAB 
According to your suggestion, I added the command before yolo layer. but it was not working. However, I modified my code in network.c and parser.c, like this,
I added this line in network.c  `if(l.stopbackward) break;`
I added this line in parser.c  'l.stopbackward = option_find_int_quiet(options, "stopbackward", 0);'
After modification, when I want to build it using "make" , it gives the following error:
"error: 'layer {aka struct layer}' has no member named 'stopbackward'
         if(l.stopbackward) break; "
Would you give some advice please?

N. B. I am using yolov2 now

Thanks

**Comment by AlexeyAB** [comment]
> I added this line in network.c if(l.stopbackward) break;

Why do you need to add it, while it is already there? https://github.com/AlexeyAB/darknet/blob/9c9344a1ff4541499c8f69ea997be5145b4e1de3/src/network.c#L319

**Comment by ibrahimLearning** [comment]
@AlexeyAB Thank you very much, I understood my problem.  
However,
In my detection, the rectangle is bigger than object!! Would you tell me please what parameters are responsible for that issue, only anchor or others also?

N. B. I generated anchor from using my training data as per yolo instruction

**Comment by AlexeyAB** [comment]
@ibrahimLearning 

> Would you tell me please what parameters are responsible for that issue, only anchor or others also?

Yolo draws the same rectangles as in your Training dataset, so decrease size of truth in your training dataset.

Also anchors affects on object size.

**Comment by ibrahimLearning** [comment]
@AlexeyAB 
Thank you very much for your all answers. I have one more question,

Which code segments are responsible for "backward" command, such as in the detector.c, network.c, parser.c, etc?

It will be easier for me to modify it if necessary. 
Thank you very much for your help.
